<img src="https://www.uc3m.es/ss/Satellite?blobcol=urldata&blobkey=id&blobtable=MungoBlobs&blobwhere=1371573952659">

---

# WEB ANALYTICS COURSE 4 - SEMESTER 2
# BACHELOR IN DATA SCIENCE AND ENGINEERING

# LAB 1.2 WEB SCRAPING WITH SCRAPY


---

### Marina Gómez Rey (100472836)
### María Ángeles Magro Garrote (100472867)
### Sara Piñas García (100472784)
### Eduardo González Agüero (100472704)

# 0. Lab Preparation

1.  Study and have clear the concepts explained in the theoretical class and the introductory lab.

2.   Gain experience with the use of the [Scrapy](https://scrapy.org/). The exercises of this lab will be mainly based on the utilization of functions offered by this library.

3. It is assumed students have experience in using Python notebooks. Either a local installation (e.g., local python installation + Jupyter) or a cloud-based solution (e.g., Google Colab). *We recommend the second option*.

# 1. Lab Introduction

* In this lab, we will implement a web scraper using [Scrapy](https://scrapy.org/). One of the tools explained in the theoretical class.

* The lab will be done in groups of 4 people.

* The lab defines a set of milestones the students must complete. Upon completing all the milestones, students should call the professor, who will check the correctness of the solution (*If the professor is busy, do not wait for them, move to the next lab*).

* **The final mark will be computed as a function of the number of milestones successfully completed.**

* **Each group should also share their lab notebook with the professor upon the finalization of the lab.**

* In this lab we will use the [Scrapy](https://scrapy.org/) library for the creation of a web scraper, to extract information from the web. As indicated in the *Lab Preparation* section above, it is expected that students have gained experience in the use of the library before starting the first session of the lab.

- It is recommended to use [Google Colab](https://colab.research.google.com/) to produce the Python notebook with the solution of the lab. Of course, if any student prefers using its local programming environment (e.g., jupyter) and python installation, they are welcome to do so.

In [1]:
%%capture
!!pip install scrapy

# MILESTONE 1

a) Create a _Scrapy_ project for this lab (**HINT:** Remember you can use the command `startproject`).

b) Create a crawler/spider to the website [BACHELOR IN DATA SCIENCE AND ENGINEERING
](https://www.uc3m.es/bachelor-degree/data-science).

c) Add the code to the crawler to get PROGRAM header. **TIP:** Find the element tag with `id="program"` and print the result.

d) Add the code to the crawler to find the table inside PROGRAM for Course 1 - Semester 1 and print the result.

In [2]:
%mkdir "Lab2_scrapy"
%cd "Lab2_scrapy"

/content/Lab2_scrapy


In [3]:
import scrapy
!scrapy startproject lab_scrapy_1_2

New Scrapy project 'lab_scrapy_1_2', using template directory '/usr/local/lib/python3.10/dist-packages/scrapy/templates/project', created in:
    /content/Lab2_scrapy/lab_scrapy_1_2

You can start your first spider with:
    cd lab_scrapy_1_2
    scrapy genspider example example.com


In [4]:
%cd "lab_scrapy_1_2"

/content/Lab2_scrapy/lab_scrapy_1_2


In [5]:
!scrapy genspider data_spider https://www.uc3m.es/bachelor-degree/data-science

Created spider 'data_spider' using template 'basic' in module:
  lab_scrapy_1_2.spiders.data_spider


    def parse(self, response):
        print(response.xpath("//*[contains(text(),'Program')]").get())


In [7]:
!scrapy crawl data_spider

2024-09-25 09:12:07 [scrapy.utils.log] INFO: Scrapy 2.11.2 started (bot: lab_scrapy_1_2)
2024-09-25 09:12:07 [scrapy.utils.log] INFO: Versions: lxml 4.9.4.0, libxml2 2.10.3, cssselect 1.2.0, parsel 1.9.1, w3lib 2.2.1, Twisted 24.7.0, Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0], pyOpenSSL 24.2.1 (OpenSSL 3.3.2 3 Sep 2024), cryptography 43.0.1, Platform Linux-6.1.85+-x86_64-with-glibc2.35
2024-09-25 09:12:07 [scrapy.addons] INFO: Enabled addons:
[]
2024-09-25 09:12:07 [asyncio] DEBUG: Using selector: EpollSelector
2024-09-25 09:12:07 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
2024-09-25 09:12:07 [scrapy.utils.log] DEBUG: Using asyncio event loop: asyncio.unix_events._UnixSelectorEventLoop
2024-09-25 09:12:07 [scrapy.extensions.telnet] INFO: Telnet Password: 53abfa806a8e78d5
2024-09-25 09:12:07 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',


    def parse(self, response):
        program_section = response.xpath('.//h3[contains(text(), "Year 1 - Semester 1")]/following-sibling::table//tr')

        for row in program_section:
            cells = row.xpath('.//td//text()').getall()
            print(cells)

In [17]:
!scrapy crawl data_spider

2024-09-25 09:38:47 [scrapy.utils.log] INFO: Scrapy 2.11.2 started (bot: lab_scrapy_1_2)
2024-09-25 09:38:47 [scrapy.utils.log] INFO: Versions: lxml 4.9.4.0, libxml2 2.10.3, cssselect 1.2.0, parsel 1.9.1, w3lib 2.2.1, Twisted 24.7.0, Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0], pyOpenSSL 24.2.1 (OpenSSL 3.3.2 3 Sep 2024), cryptography 43.0.1, Platform Linux-6.1.85+-x86_64-with-glibc2.35
2024-09-25 09:38:47 [scrapy.addons] INFO: Enabled addons:
[]
2024-09-25 09:38:47 [asyncio] DEBUG: Using selector: EpollSelector
2024-09-25 09:38:47 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
2024-09-25 09:38:47 [scrapy.utils.log] DEBUG: Using asyncio event loop: asyncio.unix_events._UnixSelectorEventLoop
2024-09-25 09:38:47 [scrapy.extensions.telnet] INFO: Telnet Password: 594e27fef50a0df0
2024-09-25 09:38:47 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',

# MILESTONE 2

a) Obtain the link to Web Analytics course by finding the corresponding href.

b) Create a new spider _class_ and access to this URL.

**TIP**: For this milestone, you need to create a new crawler and give it a different name.

c) Print the text inside the _Description of contents: programme_ section.



    def parse(self, response):
        web_analytics = response.xpath('//a[span[contains(text(), "Web Analytics")]]/@href').get()
        print(web_analytics)

In [20]:
!scrapy crawl data_spider

2024-09-25 09:46:58 [scrapy.utils.log] INFO: Scrapy 2.11.2 started (bot: lab_scrapy_1_2)
2024-09-25 09:46:58 [scrapy.utils.log] INFO: Versions: lxml 4.9.4.0, libxml2 2.10.3, cssselect 1.2.0, parsel 1.9.1, w3lib 2.2.1, Twisted 24.7.0, Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0], pyOpenSSL 24.2.1 (OpenSSL 3.3.2 3 Sep 2024), cryptography 43.0.1, Platform Linux-6.1.85+-x86_64-with-glibc2.35
2024-09-25 09:46:58 [scrapy.addons] INFO: Enabled addons:
[]
2024-09-25 09:46:58 [asyncio] DEBUG: Using selector: EpollSelector
2024-09-25 09:46:58 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
2024-09-25 09:46:58 [scrapy.utils.log] DEBUG: Using asyncio event loop: asyncio.unix_events._UnixSelectorEventLoop
2024-09-25 09:46:58 [scrapy.extensions.telnet] INFO: Telnet Password: f7ba20e200c93849
2024-09-25 09:46:58 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',

In [21]:
!scrapy genspider web_spider https://aplicaciones.uc3m.es/cpa/generaFicha?&est=350&plan=392&asig=16507&idioma=2

Created spider 'web_spider' using template 'basic' in module:
  lab_scrapy_1_2.spiders.web_spider


In [22]:
!scrapy crawl web_spider

2024-09-25 09:49:22 [scrapy.utils.log] INFO: Scrapy 2.11.2 started (bot: lab_scrapy_1_2)
2024-09-25 09:49:22 [scrapy.utils.log] INFO: Versions: lxml 4.9.4.0, libxml2 2.10.3, cssselect 1.2.0, parsel 1.9.1, w3lib 2.2.1, Twisted 24.7.0, Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0], pyOpenSSL 24.2.1 (OpenSSL 3.3.2 3 Sep 2024), cryptography 43.0.1, Platform Linux-6.1.85+-x86_64-with-glibc2.35
2024-09-25 09:49:22 [scrapy.addons] INFO: Enabled addons:
[]
2024-09-25 09:49:22 [asyncio] DEBUG: Using selector: EpollSelector
2024-09-25 09:49:22 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
2024-09-25 09:49:22 [scrapy.utils.log] DEBUG: Using asyncio event loop: asyncio.unix_events._UnixSelectorEventLoop
2024-09-25 09:49:22 [scrapy.extensions.telnet] INFO: Telnet Password: e3f1fdeae2a7153f
2024-09-25 09:49:22 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',


    def parse(self, response):
        programme = response.xpath('.//*[contains(text(), "Description of contents: programme")]/following-sibling::div').xpath('string(.)').get()
        print(programme)

In [43]:
!scrapy crawl web_spider

2024-09-25 10:35:48 [scrapy.utils.log] INFO: Scrapy 2.11.2 started (bot: lab_scrapy_1_2)
2024-09-25 10:35:48 [scrapy.utils.log] INFO: Versions: lxml 4.9.4.0, libxml2 2.10.3, cssselect 1.2.0, parsel 1.9.1, w3lib 2.2.1, Twisted 24.7.0, Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0], pyOpenSSL 24.2.1 (OpenSSL 3.3.2 3 Sep 2024), cryptography 43.0.1, Platform Linux-6.1.85+-x86_64-with-glibc2.35
2024-09-25 10:35:48 [scrapy.addons] INFO: Enabled addons:
[]
2024-09-25 10:35:48 [asyncio] DEBUG: Using selector: EpollSelector
2024-09-25 10:35:48 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
2024-09-25 10:35:48 [scrapy.utils.log] DEBUG: Using asyncio event loop: asyncio.unix_events._UnixSelectorEventLoop
2024-09-25 10:35:48 [scrapy.extensions.telnet] INFO: Telnet Password: b49c57a7345fb6fb
2024-09-25 10:35:48 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',

# MILESTONE 3

a) Modify your code from previous milestones for running both crawlers in the same command line process.


b) Instead of printing the results (from Milestone 1 and 2), save them in a file.

In [47]:
!scrapy genspider total_spider https://www.uc3m.es/

Spider 'total_spider' already exists in module:
  lab_scrapy_1_2.spiders.total_spider


import scrapy
from scrapy.crawler import CrawlerProcess
from scrapy.utils.project import get_project_settings

class BachelorSpider(scrapy.Spider):
    name = "bachelor_spider"
    allowed_domains = ["www.uc3m.es"]
    start_urls = ["https://www.uc3m.es/bachelor-degree/data-science"]

    def parse(self, response):
        program_section = response.xpath('.//h3[contains(text(), "Year 1 - Semester 1")]/following-sibling::table//tr')
        results = []

        for row in program_section:
            cells = row.xpath('.//td//text()').getall()
            results.append(' '.join(cells))
        
        with open('results.txt', 'a') as f:
            for result in results:
                f.write(result + '\n')

class WebSpider(scrapy.Spider):
    name = "web_spider"
    allowed_domains = ["aplicaciones.uc3m.es"]
    start_urls = ["https://aplicaciones.uc3m.es/cpa/generaFicha?&est=350&plan=392&asig=16507&idioma=2"]

    def parse(self, response):
        programme = response.xpath('.//*[contains(text(), "Description of contents: programme")]/following-sibling::div').xpath('string(.)').get()
        
        with open('results.txt', 'a') as f:  
            f.write(f"Programme Description: {programme}\n")  

settings = get_project_settings()
process = CrawlerProcess(settings)
process.crawl(BachelorSpider)
process.crawl(WebSpider)
process.start()

In [48]:
!scrapy crawl total_spider

/usr/local/lib/python3.10/dist-packages/scrapy/spiderloader.py:49: UserWarning: There are several spiders with the same name:

  WebSpider named 'web_spider' (in lab_scrapy_1_2.spiders.total_spider)

  WebSpiderSpider named 'web_spider' (in lab_scrapy_1_2.spiders.web_spider)

  This can cause unexpected behavior.
  warnings.warn(
2024-09-25 10:38:39 [scrapy.utils.log] INFO: Scrapy 2.11.2 started (bot: lab_scrapy_1_2)
2024-09-25 10:38:39 [scrapy.utils.log] INFO: Versions: lxml 4.9.4.0, libxml2 2.10.3, cssselect 1.2.0, parsel 1.9.1, w3lib 2.2.1, Twisted 24.7.0, Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0], pyOpenSSL 24.2.1 (OpenSSL 3.3.2 3 Sep 2024), cryptography 43.0.1, Platform Linux-6.1.85+-x86_64-with-glibc2.35
2024-09-25 10:38:39 [scrapy.addons] INFO: Enabled addons:
[]
2024-09-25 10:38:39 [asyncio] DEBUG: Using selector: EpollSelector
2024-09-25 10:38:39 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
2024-09-25 10:38:39 

**Results.txt:**

Calculus I 6 BC

Introduction to Data Science 6 BC

Linear algebra 6 BC

Probability and Data Analysis 6 BC

Programming 6 BC

Programme Description:

1.	Data Collection in the Web ecosystem:
        
        1.1  Scrapers, Crawlers
        
        1.2	APIs
2.	Data Analytics in the web

        2.1  Graph Analysis: Centrality and Influence metrics

        2.2  Network structure:
      
        2.2.1 Type of networks (bipartite graph, small world, scale free)
               
      2.2.2 Clustering, Community Detection, K-core decomposition

3.	Web data visualization

        3.1	Representation of web information.

        3.2	Visualization tools.

4.	Final Web Analytics Project

        4.1	The project needs to include the three components presented above (Data Collection, Data Analytics and Data Visualization

        
